server/controllers/annotation_controller.js

```diff
@@ -87,3 +88,13 @@ export const editAnnotation = (userId, annotationId, updateText) => {
   const update = { $set: { text: updateText, editDate: Date.now(), edited: true } };
   return Annotation.findOneAndUpdate(conditions, update, { new: true });
 };
+
+export const deleteAnnotation = (annotationId) => {
+  return Annotation.find({ parent: annotationId, deleted: false }).then(replies => {
+    if (replies.length > 0) {
+      return Annotation.findOneAndUpdate({ _id: annotationId }, { $set: { deleted: true } }, { new: true });
+    } else {
```

Comment:
Issue with this is that even if all children are deleted, there could still be non-deleted descendants (children of deleted comments).
If we assume that deleted "leaf" comments are always removed from the database, then each existing "virtually" deleted child must have some nondeleted descendant, and thus we only remove this comment if it has no children (i.e. it itself is a leaf). To maintain that invariant, when we remove a leaf, we have to check its parent, and if the parent is also deleted and is now a leaf, we delete it as well (and continue the process recursively). This can be done without searching for the parent's children if annotations at least keep track of the number of their direct children (including virtually deleted children).